clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java

```diff
@@ -127,6 +128,24 @@ public void awaitMetadataUpdate() {
     }
 
     /**
+     * Block until the metadata has been refreshed or the timeout has expired.
+     * @param timeout The maximum duration (in ms) to wait for the request
+     * @throws TimeoutException if the timeout has expired
+     */
+    public void awaitMetadataUpdate(long timeout) {
+        long now = time.milliseconds();
+        long deadline = now + timeout;
+        int version = this.metadata.requestUpdate();
+        while (this.metadata.version() == version && now < deadline) {
```

Comment:
The tough part about the consumer is that the handling is context-dependent. If the call to awaitMetadataUpdate happens in KafkaConsumer.poll(), we might want to ignore the timeout and just return an empty record set to the user. If it's a blocking call, we want to throw the exception. I wonder if it would be a good idea to let this variant return a boolean indicating whether the update happened so that the caller can more easily choose what to do?